File: cpp/00020.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> complement = {{'{', '}'}, {'(', ')'}, {'[', ']'}};
        stack<char> stk;
        
        for (int i = 0; i != s.size(); i++)
            if (complement.find(s[i]) != complement.end())
                stk.push(complement[s[i]]);
            else if (!stk.empty() && s[i] == stk.top())
                stk.pop();
            else return false;
        
        return stk.empty();
    }
};

File: python/00020.py
class Solution:
    def isValid(self, s: str) -> bool:
        mapping = { "(": ")", "[": "]", "{": "}" }
        stack = []
        for c in s:
            if c in mapping:
                stack.append(mapping[c])
            elif stack and stack[-1] == c:
                stack.pop()
            else:
                return False
                
        return len(stack) == 0

